### PR TITLE
Allows adding properties to the span

### PR DESCRIPTION
**Pull request by azanin**
Provides `reqwest-tracing` with a `trait` `ReqwestOtelSpanBackend`, allowing custom fields/behaviour to be defined.

This is a breaking change as the `TracingMiddleware` signature now requires you to define a `ReqwestOtelSpanBackend` to provide it with the desired span behaviour. 

Example:
```rust
pub struct TimeTrace;

impl ReqwestOtelSpanBackend for TimeTrace {
    fn on_request_start(req: &Request, extension: &mut Extensions) -> Span {
        extension.insert(Instant::now());
        reqwest_otel_span!(req, time_elapsed = tracing::field::Empty)
    }

    fn on_request_end(span: &Span, outcome: &Result<Response>, extension: &mut Extensions) {
        let time_elapsed = extension.get::<Instant>().unwrap().elapsed().as_millis() as i64;
        default_on_request_end(span, outcome);
        span.record("time_elapsed", &time_elapsed);
    }
}

let client = ClientBuilder::new(reqwest::Client::new())
        .with(TracingMiddleware::<TimeTrace>::new())
        .build();
```

Contributors:
@tl-flavio-barinas 
@azanin

### CODE REVIEW

File: reqwest-tracing/CHANGELOG.md
Comment (LukeMathWalker) [inline reqwest-tracing/CHANGELOG.md:12]: Was this released? Otherwise we need to cut a new patch before breaking.

File: reqwest-tracing/src/reqwest_otel_span_macro.rs
Comment (LukeMathWalker) [reply reqwest-tracing/src/reqwest_otel_span_macro.rs:158]: If the fields are not there, the `record` call will simply desugar to a no-op. I don't think that's an issue.

File: reqwest-tracing/CHANGELOG.md
Comment (LukeMathWalker) [reply reqwest-tracing/CHANGELOG.md:12]: Let's cut a release in a separate PR and we can merge this after that once it's ready.

File: reqwest-tracing/src/reqwest_otel_span_builder.rs
Comment (LukeMathWalker) [reply reqwest-tracing/src/reqwest_otel_span_builder.rs:28]: I am unconvinced - this can be done in a single handler that matches on the outcome.
If you want to override the behaviour just for successes, you can do:

```
fn on_request_end(...) {
   if let Err(error) = outcome {
       // custom logic
   } else {
       // invoke default handler
   }
}
```

File: reqwest-tracing/src/reqwest_otel_span_macro.rs
Comment (tl-flavio-barinas) [reply reqwest-tracing/src/reqwest_otel_span_macro.rs:158]: Suppose an implementor defines a field with a default value and a name the default implementation writes to. In that case, it could be an issue because the default implementation will overwrite it.

Comment (LukeMathWalker) [reply reqwest-tracing/src/reqwest_otel_span_macro.rs:158]: I see. I don't see the need for a macro though - we can simply provide default function handlers.

**Comment by azanin** [comment]
Some of the file changes are related to auto-format. Should we revert the changes in `reqwest-retry` and `reqwest-middleware`?

File: reqwest-retry/Cargo.toml
Comment (azanin) [inline reqwest-retry/Cargo.toml:13]: This is because we want to release it after merging this. We will set to the next development version in a following PR

File: reqwest-tracing/src/reqwest_otel_span_builder.rs
Comment (LukeMathWalker) [inline reqwest-tracing/src/reqwest_otel_span_builder.rs:52]: Can we extract these field names in constants?

File: README.md
Comment (LukeMathWalker) [inline README.md:45]: I'd suggest to use the fully qualified syntax here to make it easier for users to then adapt the invocation when they need custom spans - i.e. `TracingMiddleware::< DefaultSpanBackend>::new()`

Comment (azanin) [reply README.md:45]: There is already an example of a custom span in the `reqwest-tracing` README.
https://github.com/TrueLayer/reqwest-middleware/pull/43/files#diff-4ef5ea57cfc232174dcd3985f9f2aca2f3dd55cb90ae786019a8bfda2d7666faR68

**Review by tl-flavio-barinas — APPROVED** [review]

**Pull request merged**
